### DISCUSSION TITLE
How we can use Elsa hangfire or Elsa  quartz  in mongo db ?

### DISCUSSION THREAD

**Discussion by dwevedivaibhav** [elsa-workflows/elsa-core#3436] [Q&A]
@sfmskywalker,

I wanted to use Elsa Hangfire or Elsa quartz to configure for mongo DB when server will scale , I am facing difficulty when new instance is coming and previous server instance getting shut down ,Workflow not able to trigger/start again from where it got finished from previous one.

any leads it will be appreciated .

Thanks

**Comment by FransVanEk** [comment]
Can you elaborate a bit more about the objectives you want to achieve?

**Reply by dwevedivaibhav** [reply]
Hi @FransVanEk ,

Thank for responding my Question, I have two query while scaling the server and current server got crash

1) I wanted to achieve the workflow recovery while scaling the server ( for an example : I am executing one send message custom activity and having 1000 number for process  , 500 number we have sent the message and server got crashed now in current activity how can i resume the send message activity which actually start from 501 not from 1)

2) I am using mongo DB , i wanted to enable the cluster mode of operation for mongo db in Quartz , which I am struggling to configure that
any leads it will be appreciated 

Thanks,

**Reply by FransVanEk** [reply]
What is the current workflow instance? I am a bit confused about which server crashed. and which state each activity is in. 

Where do you keep track of the numbers that have been send? is this a variable in the workflow?
 
Quartz is a timing element. do you want to schedule that activity? 

can you share the workflows perhaps?

**Reply by dwevedivaibhav** [reply]
Hi @FransVanEk ,
i am just giving the example if workflow is running when you deployed in server and instance got crashed due to high CPU utilization or memory issue and new instance came up in this case how i can resume my workflow from where it got crashed and pick the right index value .  
For testing i am starting the workflow in local and running Long running process activity executes for a short while, kill the server, start the server again, and observing that the workflow engine picks up where it left off, but it picking from start index 0 even through i am setting the index value in variable  after starting the again i am seeing index value zero in variable . Do you have idea how to save immediate index value after initialization in activity?  

(Eq : (ActivityExecutionContext) context.SetVariable("CustomActivityIndex", i +1)) its showing that variable got set but its not saving in db, because after kill the server and starting it showing again 0


Yes i wanted to schedule the activity from Quartz  and wanted to use like below code:

services.AddElsa(elsa => elsa
    .AddQuartzTemporalActivities(configureQuartz: quartz => quartz.UsePersistentStore(store =>
    {
        store.UseJsonSerializer();
        store.UseMongoDBServer("Server=local;Database=Elsa;");
        store.UseClustering();
    }));

but looks like quartz not supporting for mongo db .

**Comment by FransVanEk** [comment]
OK, for the first part, it seems the activity has too much responsibility to make proper use of the capabilities of the engine. 

**Do you have idea how to save immediate index value after initialization in activity?**
No, there is no way because the engine was designed differently. 

Reasoning:
The data gets persistent in the workflow when the activity is finished. So setting it multiple times will not work. 
It seems logical to make a process that iterates over the instances that need to be handled and let the process deal with that. then when the server crashes, the workflow holds a state about the progress, which can continue from where it stopped. For that to happen, the process needs to have a loop, and after executing one item, the variable is set so the process knows which is next in line. 

As for the other question have no idea, I have never worked with it. 

Some others have, but they offer support through the discord channel. Perhaps asking about that part of your issue might get you the help you need. 

https://discord.com/channels/814605913783795763/920237171607224340

**Reply by dwevedivaibhav** [reply]
Thanks for helping , let me check the link which you shared
